standalone/pruntime/enclave/src/libc_hacks.rs

```diff
@@ -0,0 +1,424 @@
+use libc::{self, c_char, c_int, c_uint, c_ulong, c_void, iovec, size_t, ssize_t};
+use sgx_libc::{self, ocall};
+use sgx_types::{sgx_read_rand, sgx_status_t};
+use sgx_unwind as _;
+use std::{
+    ffi::CStr,
+    sync::atomic::{AtomicU16, Ordering},
+};
+
+macro_rules! assert_eq_size {
+    ($x:ty, $($xs:ty),+ $(,)?) => {
+        const _: fn() = || {
+            $(let _ = std::mem::transmute::<$x, $xs>;)+
+        };
+    };
+}
+
+macro_rules! function {
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
+        &name[..name.len() - 3]
+    }};
+}
+
+macro_rules! not_allowed {
+    () => {
+        not_allowed!(-1)
+    };
+    ($rv: expr) => {{
+        error!("NOT ALLOED TO CALL {}", function!());
+        sgx_libc::set_errno(libc::EPERM);
+        ($rv) as _
+    }};
+}
+
+#[no_mangle]
+pub extern "C" fn posix_memalign(memptr: *mut *mut c_void, align: size_t, size: size_t) -> c_int {
+    unsafe {
+        let ptr = sgx_libc::memalign(align, size);
+        // man posix_memalign:
+        // posix_memalign() returns zero on success, or one of the error values listed in the next section on failure.  The value of errno is not set.  On Linux (and other systems),
+        // posix_memalign() does not modify memptr on failure.  A requirement standardizing this behavior was added in POSIX.1-2016.
+        if ptr.is_null() && size > 0 {
+            libc::ENOMEM
+        } else {
+            *memptr = ptr;
+            0
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn read(fd: c_int, buf: *mut c_void, count: size_t) -> ssize_t {
+    if fd != 0 {
+        not_allowed!()
+    } else {
+        unsafe { ocall::read(fd, buf, count) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn readv(fd: c_int, iov: *const iovec, iovcnt: c_int) -> ssize_t {
+    assert_eq_size!(libc::iovec, sgx_libc::iovec);
+
+    if fd != 0 {
+        not_allowed!()
+    } else {
+        unsafe { ocall::readv(fd, iov as _, iovcnt) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn write(fd: c_int, buf: *const c_void, count: size_t) -> ssize_t {
+    if !matches!(fd, 1 | 2) {
+        not_allowed!()
+    } else {
+        unsafe { ocall::write(fd, buf, count) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn writev(fd: c_int, iov: *const iovec, iovcnt: c_int) -> ssize_t {
+    assert_eq_size!(libc::iovec, sgx_libc::iovec);
+
+    if !matches!(fd, 1 | 2) {
+        not_allowed!()
+    } else {
+        unsafe { ocall::writev(fd, iov as _, iovcnt) }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn getcwd(buf: *mut c_char, size: size_t) -> *mut c_char {
+    // Enclave have no working directory, let's return ""
```

Comment:
Is it better to return `/`?

Reply:
`/` is a valid path. It seems wired to return it.

Reply:
After read the doc of getcwd, wo get a better choice: `(unreachable)`

```
       If the current directory is not below the root directory of the current process (e.g., because the process set a new filesystem root  using  chroot(2)  without
       changing its current directory into the new root), then, since Linux 2.6.36, the returned path will be prefixed with the string "(unreachable)".  Such behavior
       can also be caused by an unprivileged user by changing the current directory into another mount namespace.  When dealing with  paths  from  untrusted  sources,
       callers of these functions should consider checking whether the returned path starts with '/' or '(' to avoid misinterpreting an unreachable path as a relative
       path
```

Reply:
Looks good